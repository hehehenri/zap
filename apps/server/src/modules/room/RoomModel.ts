import { Document, Schema, Types, model } from "mongoose";
import { MessageDocument } from "../message/MessageModel";

export interface RoomDocument extends Document {
  participants: Types.ObjectId[];
  createdAt: Date,
  lastMessage: MessageDocument | null,
  updatedAt: Date,
}

const roomSchema = new Schema<RoomDocument>(
  {
    participants: [{ type: Schema.Types.ObjectId, ref: "User", required: true }],
    lastMessage: { type: Schema.Types.ObjectId, ref: "Message", required: false, default: null }
  },
  {
    collection: "Room",
    timestamps: true
  }
)


export const RoomModel = model<RoomDocument>("Room", roomSchema);
