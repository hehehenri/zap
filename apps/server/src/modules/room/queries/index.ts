import { Room } from "./Room";
import { Rooms } from "./Rooms";

export const roomQueries = {
  rooms: Rooms,
  room: Room 
}
