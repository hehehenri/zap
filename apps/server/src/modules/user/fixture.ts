import config from "../../config";
import { UserDefinition, UserModel } from "./UserModel"
import bcrypt from "bcryptjs";

let counter = 0;

const getCounter = () => {
  counter += 1;

  return counter;
}

type Fields = Pick<UserDefinition, Exclude<keyof UserDefinition, "_id">>

export const createUser = async (fields?: DeepPartial<Fields>) => {
  const index = getCounter();

  const plainTextPassword = fields?.password ?? "password";

  const password = bcrypt.hashSync(plainTextPassword, config.jwt.saltRounds);
  const username = fields?.username ?? `user#${index}`;

  const user = new UserModel({ password, username });
  await user.save();

  return user;
}
