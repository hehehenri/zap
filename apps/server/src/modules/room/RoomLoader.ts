import { createLoader } from "@entria/graphql-mongo-helpers";
import { RoomModel } from "./RoomModel";

export const RoomLoader = createLoader({
  model: RoomModel,
  loaderName: 'RoomLoader',
  defaultSort: {
    lastMessage: -1, "lastMessage.createdAt": 1
  }
})
