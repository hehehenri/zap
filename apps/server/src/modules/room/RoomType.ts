import { GraphQLNonNull, GraphQLObjectType, GraphQLString, GraphQLList, GraphQLNullableType } from "graphql/type";
import { connectionDefinitions } from "graphql-relay";

import { RoomDefinition } from "./RoomModel";
import UserType from "../user/UserType";
import { MessageType } from "../message/MessageType";

const list = <T extends GraphQLNullableType>(type: T) => {
  return new GraphQLNonNull(new GraphQLList( new GraphQLNonNull(type)))
}

const RoomType = new GraphQLObjectType<RoomDefinition>({
  name: "Room",
  fields: () => ({
    id: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (room) => room._id
    },
    participants: {
      type: list(UserType),
      resolve: (room) => room.participants
    },
    messages: {
      type: list(MessageType),
      resolve: (room) => room.messages
    },
    createdAt: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (room) => room.createdAt.toString(),
    },
    updatedAt: {
      type: new GraphQLNonNull(GraphQLString),
      resolve: (room) => room.updatedAt.toString(),
    }
  })
})

export const RoomConnection = connectionDefinitions({
  name: "Room",
  nodeType: RoomType
});

export default RoomType;
