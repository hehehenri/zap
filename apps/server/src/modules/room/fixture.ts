import { RoomDefinition, RoomModel } from "./RoomModel";

type Fields = Pick<RoomDefinition, Exclude<keyof RoomDefinition, "_id">>

export const createRoom = async (fields?: DeepPartial<Fields>) => {
  return await new RoomModel(fields).save();
}
