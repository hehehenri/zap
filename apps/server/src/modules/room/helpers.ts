import { UserDefinition } from "../user/UserModel";
import { RoomDefinition } from "./RoomModel";

export const isRoomMember = (room: RoomDefinition, user: UserDefinition) => {
  return room
    .participants
    .map(u => u._id.toString())
    .includes(user._id.toString());
}
