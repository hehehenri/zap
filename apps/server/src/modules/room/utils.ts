import { UserDefinition } from "../user/UserModel";
import { RoomDefinition } from "./RoomModel";

export const isParticipant = (room: RoomDefinition, user: UserDefinition) => {
  return !! (
    room.participants
    .filter(partId => partId.toString() == user._id.toString())
    .length
  );
}
