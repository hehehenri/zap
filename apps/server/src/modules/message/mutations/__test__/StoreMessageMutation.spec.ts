import { generateToken } from "../../../../authentication";
import { describeWithDb, testQuery } from "../../../../test/helpers";
import { createRoom } from "../../../room/fixture";
import { createUser } from "../../../user/fixture";

const query = ({ content, roomId }: { content?: string, roomId: string }) => ({
  query: `
    mutation StoreMessage($content: String! $roomId: ID!) {
      storeMessage(input: { content: $content roomId: $roomId}) {
        message { content }
      }
    }
  `,
  variables: {
    content: content ?? 'message content',
    roomId,
  },
});

describeWithDb("message/mutations/storeMessage", () => {
  it("should add message to room", async () => {
    const userA = await createUser({ username: 'usera' });
    const userB = await createUser({ username: 'userb' });

    const roomAB = await createRoom({ participants: [userA, userB] })
    const roomId = roomAB._id.toString();

    const token = generateToken(userA);
    const response = await testQuery({ query: query({ roomId }), token });

    expect(response.status).toBe(200);
    expect(response.body.data).toMatchSnapshot();
  });

  it("can't add message to room when not a member", async () => {
    const userA = await createUser({ username: 'usera' });
    const userB = await createUser({ username: 'userb' });
    const userC = await createUser({ username: 'userc' });

    const roomBC = await createRoom({ participants: [userB, userC] })
    const roomId = roomBC._id.toString();

    const token = generateToken(userA);
    const response = await testQuery({ query: query({ roomId }), token });

    expect(response.status).toBe(422);
    expect(response.body.errors).toMatchSnapshot();
  });
})
