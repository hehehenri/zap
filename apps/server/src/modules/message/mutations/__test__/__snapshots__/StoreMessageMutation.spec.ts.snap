// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message/mutations/storeMessage can't add message to room when not a member 1`] = `
[
  {
    "locations": [
      {
        "column": 7,
        "line": 3,
      },
    ],
    "message": "you can't send messages on this room",
    "path": [
      "storeMessage",
    ],
  },
]
`;

exports[`message/mutations/storeMessage should add message to room 1`] = `
{
  "storeMessage": {
    "message": {
      "content": "message content",
    },
  },
}
`;
